apm-sniffer/optional-plugins/armeria-0.85.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/armeria/ArmeriaClientInterceptor.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.skywalking.apm.plugin.armeria;
+
+import java.lang.reflect.Method;
+import java.net.URI;
+
+import com.linecorp.armeria.client.Clients;
+import com.linecorp.armeria.common.HttpHeadersBuilder;
+import com.linecorp.armeria.common.HttpMethod;
+import com.linecorp.armeria.common.HttpRequest;
+import com.linecorp.armeria.common.util.SafeCloseable;
+import io.netty.util.AsciiString;
+import org.apache.skywalking.apm.agent.core.context.CarrierItem;
+import org.apache.skywalking.apm.agent.core.context.ContextCarrier;
+import org.apache.skywalking.apm.agent.core.context.ContextManager;
+import org.apache.skywalking.apm.agent.core.context.tag.Tags;
+import org.apache.skywalking.apm.agent.core.context.trace.AbstractSpan;
+import org.apache.skywalking.apm.agent.core.context.trace.SpanLayer;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.EnhancedInstance;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.InstanceMethodsAroundInterceptor;
+import org.apache.skywalking.apm.network.trace.component.ComponentsDefine;
+
+/**
+ * @author kezhenxu94
+ */
+public abstract class ArmeriaClientInterceptor implements InstanceMethodsAroundInterceptor {
+    private static final String KEY_SAFE_CLOSEABLE = "SAFE_CLOSEABLE";
+
+    protected void beforeMethod(final URI uri, final HttpMethod httpMethod, final String path) {
+        final ContextCarrier contextCarrier = new ContextCarrier();
+        final String remotePeer = uri.getHost() + ":" + uri.getPort();
+
+        final AbstractSpan exitSpan = ContextManager.createExitSpan(path, contextCarrier, remotePeer);
+
+        exitSpan.setComponent(ComponentsDefine.ARMERIA);
+        exitSpan.setLayer(SpanLayer.HTTP);
+        Tags.HTTP.METHOD.set(exitSpan, httpMethod.name());
+
+        ContextManager.getRuntimeContext().put(KEY_SAFE_CLOSEABLE, Clients.withHttpHeaders(headers -> {
```

Comment:
Do you mean this? I don't open to check the source codes. Could this only be written in JDK8, right?

Reply:
Why does a header OP require to close manually?

Reply:
> Do you mean this? I don't open to check the source codes. Could this only be written in JDK8, right?

yes

> Why does a header OP require to close manually?
 
don't need to close **manually**, but should not be close **too early**, so I put the close operation in the `afterMethod`

Reply:
the recommended style is `try-with-resources`, but it's obviously not applicable here, closing the `SafeClosedble` resulting removal of the headers

Reply:
My point is not this. I mean, does this #withHttpHeader function create a new client requiring close by you? If it is the original client, the close OP is developer's responsibility, not plugin's. You may close the client unexpectedly.

Reply:
I’m not closing the client, neither “withHttpHeader” creates client, it returns a callback to clear the header variables that are stored in thread, and the “close” method just clear the headers in the thread, see documentation of Armeria https://line.github.io/armeria/client-custom-http-headers.html

Reply:
And we are obviously responsible to remove these headers that are set by us

Reply:
Exactly, thanks! @kezhenxu94
This is just called when `ClientRequestContext` is created and update the `HttpHeaders` with the new one.

Reply:
Thanks for sharing the context. Didn't expect the Armeria is using ThreadLocal to process the ClientRequestContext. Most RPC frameworks using a local variable ref.